### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by nicogbg**
Add ZigBee DSB adapter tutorial to AllJoyn sample

### CODE REVIEW

**Comment by zhuridartem** [comment]
Ping. Any update on this? This has been idle for some time now.

File: en-US/win10/samples/ZigBeeAdapterTutorial.md [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+Acronyms:
+- ZDO: ZigBee Device Object
+- ZCL: ZigBee Cluster Library
+
+## Prerequisites
+1. XBee ZigBee module from [Digi](http://www.digi.com), e.g.: XB24 Z7PIT-004
+2. XBee Explorer USB dongle from [SparkFun](https://www.sparkfun.com)
+3. [XCTU](http://www.digi.com/products/xbee-rf-solutions/xctu-software/xctu) tool from Digi
+4. Windows 10 desktop with Visual Studio 2015 and [AllJoyn Explorer (AJX)]
```
Comment (anthony-ngu) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: Please make this link on the same line. It does not generate a link otherwise.

File: en-US/win10/samples/ZigBeeAdapterTutorial.md [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+See [ZigBee standard](http://www.zigbee.org) for more information about ZigBee, ZigBee profiles, ZCL clusters...
+
+Acronyms:
+- ZDO: ZigBee Device Object
+- ZCL: ZigBee Cluster Library
+
+## Prerequisites
+1. XBee ZigBee module from [Digi](http://www.digi.com), e.g.: XB24 Z7PIT-004
+2. XBee Explorer USB dongle from [SparkFun](https://www.sparkfun.com)
```
Comment (anthony-ngu) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: Why not link to the actual XBee Explorer Dongle rather than the generic Sparkfun site?
https://www.sparkfun.com/products/11697

Or was the goal to have people look through each linked site to have to buy / find out where to buy the parts?

File: en-US/win10/samples/ZigBeeAdapterTutorial.md [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+AllJoyn Explorer and its documentation (how to install, to use…) can be find [here](http://ms-iot.github.io/content/en-US/win10/AllJoyn.htm).
+
+![ZigBeeHardware]({{site.baseurl}}/images/ZigBee/ZigBeeHardware.png)
+
+## Set up steps
+1. Install the required tools and driver listed in the prerequisites on your Windows 10 desktop
+(see their respective documentations to figure out how to proceed).
+2. Configure the XBee module
+3. Let your device join your ZigBee network
+4. set up your Raspberry Pi2 (if you target that device)
```
Comment (anthony-ngu) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: Small nit-pick: Please capitalize the S in "set"

File: en-US/win10/samples/ZigBeeAdapterTutorial.md [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+## Let ZigBee devices join your ZigBee network
+Once the XBee ZigBee module has been configured you can build your ZigBee network and let your ZigBee devices join. In order to do that you just need to power up your ZigBee devices. ZigBee Light Link (aka ZLL) and Home Automation devices will, by default and if not already part of a ZigBee network, try to join a ZigBee network which permit join is enabled. Since the XBee ZigBee module has been configured to always enable permit join, the ZigBee devices will join your network. 
+
+You can verify that devices have by using “network discovery” feature of XCTU tool.
+
+![ZigBeeJoinNetVerif]({{site.baseurl}}/images/ZigBee/ZigBeeJoinNetVerif.png)
+
+## Set up your Raspberry Pi2
+1. perform initial set up as instructed [here](http://ms-iot.github.io/content/en-US/win10/SetupRPI.htm)
```
Comment (anthony-ngu) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: Please capitalize the First letter of the first word of each of these steps to match with other documentation.

File: en-US/win10/samples/ZigBeeAdapterTutorial.md [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+ 
+ 
+### Known limitations of the current version of the ZigBee adapter
+- ZigBee adapter only discovers devices that are directly connected to the XBee module. 
+- ZigBee adapter only support some devices defined in ZigBee Light Link or Home Automation profiles. This means that it only implements the necessary ZCL clusters and ZDO commands to handle them. That said support for new devices and new ZCL clusters can easily be added to ZigBee adapter code.
+- ZigBee adapter doesn’t expose any method to commission ZigBee devices.
+
+## ZigBee adapter in detail 
+ZigBee adapter is written in C# and exposes ZigBee devices on AllJoyn through BridgeRT interface. ZigBee device is exposed in AllJoyn as follow:
+- Each ZigBee end point of a ZigBee device is exposed as an AllJoyn Service (bus attachment)
```
Comment (anthony-ngu) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md:101]: Need to add an extra line here in order to have this show up as a list

File: en-US/win10/samples/ZigBeeAdapterTutorial.md [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+4. What ParseResponse does is specific to each implementation of the ZigBeeCommand class. For example, the ParseResponse method of the DeviceAnnce class will get the 64 bit address (aka MAC address) and 16 bit address and may be more information about the signaled device and then send “device arrival” signal to BridgeRT (note that several classes are used to achieve that, see code for more). 
+
+### Creating a new ZCL cluster class
+ZigBee adapter doesn’t implement all clusters defined by ZCL. There will be cases where the ZigBee device you want to interact with won’t be supported by ZigBee adapter. In such a case you will need to add support for the missing clusters. 
+
+How to:
+
+1. You need to found out which the ZigBee Profile and ZigBee Device Category each end point of the new device supports. This can be documented by the device manufacturer or discovered by sending some ZDO command to the end points. ZigBee adapter can give you this information if you enable device discovery tracing in Logger class.
```
Comment (anthony-ngu) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: You need to find out*

File: en-US/win10/samples/ZigBeeAdapterTutorial.md [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+4. Upon reception a complete frame from the XBee module, the reception thread of the SerialController will call GetBytesFromModule callback of XBeeModule.
+5. GetBytesFromModule will parse the XBee part of the frame and check if the response received has a matching command. If so if will call the ParseResponse callback of the relevant command and then signal the reception of the response to that command. 
+
+### Receiving ZDO or ZCL command 
+ZigBee device can send ZDO or ZCL command to the XBee module, e.g.: device announce ZDO command which is sent when a device wakes up or join the network, report attribute ZCL command when an attribute of a ZCL cluster is reportable… This kind of command are not sent in response to another command and can be seen as “notification” by ZigBee adapter.  
+
+![ReceiveZdoZcl]({{site.baseurl}}/images/ZigBee/ReceiveZdoZcl.png)
+
+1. Adapter class will build a list of notification it can receive upon initialization. This list contains instance of specific ZigBeeCommand such as DeviceAnnce, ZclReportAttribute.
```
Comment (anthony-ngu) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: Adapter class will build a list of notifications*
This list contains instances*
DeviceAnnce or ZclReportAttribute*?

File: en-US/win10/samples/ZigBeeAdapterTutorial.md [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+2. Upon reception a complete frame from the XBee module, the reception thread of the SerialController will call GetBytesFromModule callback of XBeeModule.
```
Comment (anthony-ngu) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: Upon reception of a complete frame*

File: en-US/win10/samples/ZigBeeAdapterTutorial.md [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+3. GetBytesFromModule will parse the XBee part of the frame and check if it’s a response to a command that has been sent (see previous section). If not, it will go through the notification list and call the ParseResponse method of each element until 1 accept the frame or none have accepted. In that later case the frame will be thrown away.
```
Comment (anthony-ngu) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: nit-pick about 1 vs. one
Maybe reword: "In that later case the frame will be thrown away. "
To: If none have accepted, the frame will be thrown away.*

File: en-US/win10/samples/ZigBeeAdapterTutorial.md [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+Example of a new cluster class
+
+![NewClusterClass]({{site.baseurl}}/images/ZigBee/NewClusterClass.png)
+
+What needs to be done in ZclClusterFactory class
+
+![ZclFactoryChange]({{site.baseurl}}/images/ZigBee/ZclFactoryChange.png)
+
+### Quick overview of frame used by XBee API
+Digi provide API that can be used to interact with its XBee ZigBee module. This API is used to send or receive AT command or ZDO and ZCL command. AT command are specific to XBee module and can only be interpreted by XBee module whereas ZDO or ZCL command can be interpreted by any ZigBee devices. Frame used to send AT command are a bit different than frames used to send ZDO or ZCL command (see XBee ZigBee module documentation for more detail).
```
Comment (anthony-ngu) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: Digi provides an API*
AT commands are specific to the XBee module and can only be interpreted by the XBee module*
Frames used to send AT commands are a bit different than frames used to send ZDO or ZCL commands (see XBee ZigBee module documentation for more details).*

File: en-US/win10/samples/ZigBeeAdapterTutorial.md [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+
+![AJXPhilipsHue5]({{site.baseurl}}/images/ZigBee/AJXPhilipsHue5.png)
+
+### Class overview
+
+![ClassMap]({{site.baseurl}}/images/ZigBee/ClassMap.png)
+
+The __Adapter__ class is the main class of ZigBee adapter. This class derives from __IAdapter__ (BridgeRT interface) and contains a collection of __ZigBeeDevice__ instances and an instance of the __XBeeModule__ class.
+
+The __XBeeModule__ class handles the interactions with XBee module, this class has method to build and parse XBee frames. See XBee ZigBee module documentation (http://www.digi.com) for more detail about its API and its frame layout. XBeeModule class uses the __SerialController__ class that handles communication over a serial port.
```
Comment (anthony-ngu) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: I suggest changing the wording of the first sentence to:
The **XBeeModule** class handles the interactions with the XBee module and has methods to build and parse XBee frames.

Did you mean to have the "See XBee ZigBee module documentation (http://www.digi.com)" to be a link?
If so, use See [XBee ZigBee module documentation](http://www.digi.com) and probably want to change the link to the actual XBee ZigBee module documentation page.

File: en-US/win10/samples/ZigBeeAdapterTutorial.md [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+
+### Class overview
+
+![ClassMap]({{site.baseurl}}/images/ZigBee/ClassMap.png)
+
+The __Adapter__ class is the main class of ZigBee adapter. This class derives from __IAdapter__ (BridgeRT interface) and contains a collection of __ZigBeeDevice__ instances and an instance of the __XBeeModule__ class.
+
+The __XBeeModule__ class handles the interactions with XBee module, this class has method to build and parse XBee frames. See XBee ZigBee module documentation (http://www.digi.com) for more detail about its API and its frame layout. XBeeModule class uses the __SerialController__ class that handles communication over a serial port. 
+
+The __ZigBeeDevice__ class represent a ZigBee device. This class has no interface with BridgeRT since only EndPoints of a ZigBee device are expose to AllJoyn. __ZigBeeDevice__ class contains a collection of __ZigBeeEndPoint__ instances.
```
Comment (anthony-ngu) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: Suggested grammar fixes:

The **ZigBeeDevice** class represents a ZigBee device. This class has no interface with BridgeRT since only EndPoints of a ZigBee device are exposed to AllJoyn. The **ZigBeeDevice** class contains a collection of **ZigBeeEndPoint** instances.

File: en-US/win10/samples/ZigBeeAdapterTutorial.md [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+
+The __ZigBeeEndPoint__ class represent an EndPoint of a ZigBee device. This class derives from __IAdapterDevice__ (BridgeRT interface) and contains an instance of the __BasicCluster__ and a collection __ZclCluster__ instance. To be more accurate it contains a collection of cluster classes, e.g.: __OnOffCluster__, which derive from the abstract __ZclCluster__ class.
```
Comment (anthony-ngu) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: Suggested grammar fixes:

The **ZigBeeEndPoint** class represents an EndPoint of a ZigBee device*

File: en-US/win10/samples/ZigBeeAdapterTutorial.md [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+
+The cluster classes, e.g.: __BasicCluster__, __OnOffCluster__, represent a ZCL cluster. Each cluster class is a “specific” implementation of the __ZclCluster__ abstract class.
+ 
+The __ZclCluster__ class is an abstract class that derives from __IAdapterProperty__ (BridgeRT interface) and contains a collection of __ZclAttribute__ instances and a collection of __ZclCommand__ instances. A specific implementation of ZclCluster simply consists in defining the list of ZclAttribute and ZclCommand that should be supported for that specific cluster.
+
+The __ZclAttribute__ class represent an attribute as defined in ZCL standard. This class derives from __IAdapterAttribute__ (BridgeRT interface) and from __ZigBeeCommand__ class. This class contains an instance of __IAdapterValue__ class which is a BridgeRT interface. For information, reading or writing an attribute consist in sending a specific ZCL command.
```
Comment (anthony-ngu) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: Suggested grammar fixes:
The **ZclAttribute** class represents an attribute as defined in ZCL standard. This class derives from **IAdapterAttribute** (BridgeRT interface) and the **ZigBeeCommand** class.

File: en-US/win10/samples/ZigBeeAdapterTutorial.md [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+
+The __ZclCommand__ class represent a command as defined in ZCL standard. This class derives from __IAdapterMethod__ (BridgeRT interface) and from __ZigBeeCommand__ class. This class contains list of input and output parameters. These parameters are actually __IAdapterValue__ class which is a BridgeRT interface.
```
Comment (anthony-ngu) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: Suggested grammar fixes:
The **ZclCommand** class represents a command as defined in ZCL standard. This class derives from **IAdapterMethod** (BridgeRT interface) and the **ZigBeeCommand** class

File: en-US/win10/samples/ZigBeeAdapterTutorial.md [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+
+The __ZigBeeCommand__ class is an Abstract class that is used to send and receive ZDO or ZCL command. Preparing (or parsing) the ZCL or ZDO payload that XBeeModule will send to XBee module (or receive from XBeeModule) is a shared task between ZigBeeCommand class for its “generic” part and the derived class of ZigBeeCommand for the specific part. 
+
+ZDO command classes such as __ManagementLQI__ class, __ActiveEndPoints__ class… are used to discover the ZigBee network and the ZigBee devices and end points. These classes derive from __ZigBeeCommand__ class.
+
+The __ZclClusterFactory__ class is used to create instances of a specific cluster class. This class is used by the adapter class to create the relevant clusters when it discovers a new ZigBee device. Note that Adapter class create a ZigBeeDevice instance only if that device has at least 1 end point that has a ZCL cluster listed in the supported cluster list of the ZclClusterFactory.
```
Comment (anthony-ngu) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: nit-pick about 1 vs. one

File: en-US/win10/samples/ZigBeeAdapterTutorial.md [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+
+AT command classes such as __AO_Command__,  __HV_Command__…  are used by XBeeModule upon its initialization to get information about the XBee module it uses. These classes derive from the __XBeeATCommand__ abstract class. See XBee ZigBee module documentation (http://www.digi.com) for more detail about AT commands.
```
Comment (anthony-ngu) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: Did you mean to have the "See XBee ZigBee module documentation (http://www.digi.com)" to be a link?
If so, use [See XBee ZigBee module documentation](http://www.digi.com) and probably want to change the link to the actual XBee ZigBee module documentation page.

**Comment by anthony-ngu** [comment]
Very nice images and a lot of good content. Some nit-picks about grammar and a couple broken links that need fixing.

After you've fixed those issues, I can do a merge :+1:

File: en-US/win10/samples/ZigBeeAdapterTutorial.md (hunk 1 above)
Comment (nicogbg) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: fixed

File: en-US/win10/samples/ZigBeeAdapterTutorial.md (hunk 2 above)
Comment (nicogbg) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: fixed (the initial idea was to just give a link to sparkfun so if sparkfun change its web site in the future the link will remain valid)

File: en-US/win10/samples/ZigBeeAdapterTutorial.md (hunk 3 above)
Comment (nicogbg) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: fixed

File: en-US/win10/samples/ZigBeeAdapterTutorial.md (hunk 4 above)
Comment (nicogbg) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: done

File: en-US/win10/samples/ZigBeeAdapterTutorial.md (hunk 6 above)
Comment (nicogbg) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: thanks, fixed

File: en-US/win10/samples/ZigBeeAdapterTutorial.md (hunk 7 above)
Comment (nicogbg) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: typo fixed. DeviceAnnce or ZclReportAttribute are classes name which directly come from ZigBee standard

File: en-US/win10/samples/ZigBeeAdapterTutorial.md (hunk 8 above)
Comment (nicogbg) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: fixed

File: en-US/win10/samples/ZigBeeAdapterTutorial.md (hunk 9 above)
Comment (nicogbg) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: fixed

File: en-US/win10/samples/ZigBeeAdapterTutorial.md (hunk 10 above)
Comment (nicogbg) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: fixed

File: en-US/win10/samples/ZigBeeAdapterTutorial.md (hunk 11 above)
Comment (nicogbg) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: rewording done. I don't want to put a direct link to the doc since Digi changed location several times during project dev. So I'd rather point to the digi site and have the user looking for the right doc

File: en-US/win10/samples/ZigBeeAdapterTutorial.md (hunk 12 above)
Comment (nicogbg) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: done

File: en-US/win10/samples/ZigBeeAdapterTutorial.md (hunk 13 above)
Comment (nicogbg) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: done

File: en-US/win10/samples/ZigBeeAdapterTutorial.md (hunk 14 above)
Comment (nicogbg) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: done

File: en-US/win10/samples/ZigBeeAdapterTutorial.md (hunk 15 above)
Comment (nicogbg) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: done

File: en-US/win10/samples/ZigBeeAdapterTutorial.md (hunk 16 above)
Comment (nicogbg) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: done

File: en-US/win10/samples/ZigBeeAdapterTutorial.md (hunk 17 above)
Comment (nicogbg) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: done

File: en-US/win10/samples/ZigBeeAdapterTutorial.md (hunk 11 above)
Comment (nicogbg) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md]: I misunderstood your point about link => and fixed it

File: en-US/win10/samples/ZigBeeAdapterTutorial.md (hunk 5 above)
Comment (nicogbg) [inline en-US/win10/samples/ZigBeeAdapterTutorial.md:101]: I see them as a list in https://github.com/nicogbg/content/blob/develop/en-US/win10/samples/ZigBeeAdapterTutorial.md => don't know what to fix

**Comment by nicogbg** [comment]
Hi Anthony,

I've fixed the issues you pointed out...

thanks

Nicolas

**Comment by grahamehorner** [comment]
@nicogbg great work would you mind if I used this for the basis of a AllJoyn - LLAP bridge implementation?

**Comment by nicogbg** [comment]
No problem, you can definitely use it !!! 
Comment, feedback are more than welcome...

Cheers

Nicolas

**Comment by anthony-ngu** [comment]
Thanks @nicogbg. I'll merge it in now.

**Comment by nicogbg** [comment]
thanks

2015-11-02 20:02 GMT+01:00 Anthony N notifications@github.com:
